File: examples/kotlin/src/main/kotlin/adr/app/Wire.kt
// ── app/wire — THE SINGLE COMPOSITION ROOT (I1, G7) ────────────────────────
// Exactly one file may know what is real and what is faked in a build. Removing it
// means a service locator, which G7 forbids.
//
// Everything that binds is here: every port→adapter binding, the effect sink, the
// Boundary, the Controller, and the agent loop. Plugging a block in is its
// `register(...)` line plus its port binding plus its sink branch; pulling it out is
// the same list, subtracted, plus `rm -rf blocks/<X>/`.

package adr.app

import adr.blocks.analysis.AnalysisBlock
import adr.blocks.analysis.AnalysisRelay
import adr.blocks.analysis.LiveRelayWriter
import adr.blocks.artifact.ArtifactBlock
import adr.blocks.artifact.DeliveryPort
import adr.blocks.artifact.LiveDelivery
import adr.blocks.artifact.REQUEST_SEAL
import adr.blocks.console.ConsoleBlock
import adr.blocks.escalation.EscalationBlock
import adr.blocks.escalation.LivePager
import adr.blocks.escalation.OncallPort
import adr.blocks.inbox.InboxBlock
import adr.blocks.inbox.NOTE_DROP
import adr.blocks.inbox.NOTE_FAULT
import adr.blocks.triage.Ticket
import adr.blocks.triage.TriageBlock
import adr.contract.AnalysisEffect
import adr.contract.ArtifactEffect
import adr.contract.Effect
import adr.contract.EscalationEffect
import adr.contract.ToolResult
import adr.contract.TriageEffect
import adr.spine.agent.AgentLoop
import adr.spine.agent.runTurn
import adr.spine.boundary.Boundary
import adr.spine.boundary.InMemoryBus
import adr.spine.boundary.MovingClock
import adr.spine.boundary.RecordingSink
import adr.spine.boundary.SequentialIds
import adr.spine.concurrency.InMemoryRelay
import adr.spine.concurrency.SerialConsumer
import adr.spine.concurrency.TurnRunner
import adr.spine.ports.AuthorityResolver
import adr.spine.ports.Bus
import adr.spine.ports.Clock
import adr.spine.ports.ConfirmPolicy
import adr.spine.ports.EventSource
import adr.spine.ports.IdSource
import adr.spine.ports.Mailbox
import adr.spine.ports.ModelProvider
import adr.spine.ports.RelayRead
import adr.spine.ports.Sink
import adr.spine.pure.Action
import adr.spine.pure.Actor
import adr.spine.pure.Authority
import adr.spine.pure.BlockRegistration
import adr.spine.pure.ConsumerEvent
import adr.spine.pure.InputPolicy
import adr.spine.pure.KeyedEffect
import adr.spine.pure.Message
import adr.spine.pure.PerformMode
import adr.spine.pure.Registry
import adr.spine.pure.SessionId
import adr.spine.pure.Signature
import adr.spine.pure.SourceName
import adr.spine.pure.StagedInput
import adr.spine.pure.TicketId
import adr.spine.pure.rawOf
import adr.spine.pure.registryOf
import adr.spine.surface.Controller
import ai.torad.aisdk.LanguageModel

/**
 * The one branch per effect kind. Exhaustive over Effect with NO else arm, so a new
 * effect kind fails to compile until it is bound to something real (§11.1).
 *
 * REPLAY touches nothing — that is the mode's entire contract (§3.6), and it is the
 * reason a replayed trace can be driven through the SAME sink as a live one.
 */
class AppSink(
    private val oncall: OncallPort,
    private val delivery: DeliveryPort,
    private val relay: AnalysisRelay,
    private val log: MutableList<String>,
) : Sink {
    override fun perform(keyed: KeyedEffect, mode: PerformMode) {
        if (mode == PerformMode.REPLAY) return
        when (val effect = keyed.effect) {
            is Effect.Diag -> log += "diag[${effect.at.value}] ${effect.note}"

            is TriageEffect -> when (effect) {
                is TriageEffect.LogDecision ->
                    log += "priority[${effect.at.value}] ${effect.ticket.value} -> ${effect.level}" +
                        (effect.supersedes?.let { " (was $it)" } ?: "")
            }

            is EscalationEffect -> when (effect) {
                is EscalationEffect.PageOncall -> oncall.page(effect.ticket)
            }

            is ArtifactEffect -> when (effect) {
                is ArtifactEffect.DeliverArtifact -> delivery.deliver(effect.lines)
            }

            // The deep tier's publish is an ORDINARY EFFECT DESCRIPTOR (14.2), which is
            // why the tiering rung inherits replay-stubbing and RECOVERY idempotency for
            // free — the early return above already stubs it in REPLAY.
            is AnalysisEffect -> when (effect) {
                is AnalysisEffect.PublishConclusion -> relay.publish(effect.at, effect.text)
            }
        }
    }
}

/**
 * The product-owned authority seam (14.3, 17.1). `acting` is how a NON-HUMAN
 * confirmer reaches the gate: a policy tier, a second-agent reviewer or a deferred
 * approval queue sets it, the Command still stamps its Actor TRUTHFULLY, and only
 * the Authority differs — which is precisely what F3 asks the book to state plainly.
 */
class RunAuthority(
    private val agent: Authority = Authority("agent-run-7f"),
    private val human: Authority = Authority("host:marcos"),
) : AuthorityResolver {
    var acting: Authority? = null

    override fun authorityOf(by: Actor, session: SessionId): Authority = acting ?: when (by) {
        Actor.Agent -> agent
        Actor.Human -> human
    }
}

/** A product rule on top of the structural self-confirm denial the gate already applies. */
class ConfirmingAuthorities(private val allowed: Set<Authority>? = null) : ConfirmPolicy {
    override fun mayConfirm(
        sig: Signature,
        result: ToolResult,
        requestedBy: Authority,
    ): Boolean = allowed == null || sig.authority in allowed
}

/** The cognition seam, bound. The spine never names the runtime's model type; this does. */
fun modelProvider(model: LanguageModel): ModelProvider<LanguageModel> =
    object : ModelProvider<LanguageModel> {
        override fun model(): LanguageModel = model
    }

/** The sensing seam, bound: a scripted queue of untrusted perceived events (10.2). */
class ScriptedEvents(private val events: List<StagedInput.Perceived> = emptyList()) : EventSource {
    private var next = 0

    override fun poll(): StagedInput.Perceived? = events.getOrNull(next++)
}

/**
 * The fake world the offline demo and every test page, deliver and publish into.
 *
 * [store] is a constructor parameter so TWO worlds — a fast tier and a deep tier —
 * can share ONE relay while sharing nothing else: separate buses, separate sessions,
 * separate models, no handle from either to the other.
 */
class World(val store: InMemoryRelay = InMemoryRelay()) {
    val pages = mutableListOf<String>()
    val deliveries = mutableListOf<String>()
    val conclusions = mutableListOf<String>()
    val oncall: OncallPort = LivePager("https://pager.example/page") { pages += it }
    val delivery: DeliveryPort = LiveDelivery { deliveries += it }
    val relay: AnalysisRelay = LiveRelayWriter { at, text ->
        conclusions += text
        store.publish(at, text)
    }
}

data class Env(
    val clock: Clock,
    val authority: AuthorityResolver,
    val oncall: OncallPort,
    val delivery: DeliveryPort,
    val relay: AnalysisRelay,
    val ids: IdSource = SequentialIds(),
    val bus: Bus = InMemoryBus(),
    val policy: ConfirmPolicy = ConfirmingAuthorities(),
    val events: EventSource = ScriptedEvents(),
    val promptVersion: String = "triage-prompt@1",
    val session: SessionId = SessionId("session-1"),
    val tickets: List<Ticket> = emptyList(),
    /**
     * 11.4's single registry, as an ALLOWLIST. Null means every block and every verb —
     * the single-process default. [FAST_TIER] / [DEEP_TIER] are the two-tier split.
     */
    val verbs: List<BlockRegistration<State>>? = null,
    /** The barge-in rung (12), OPT-IN: no mailbox, no consumer, no cost. */
    val mailbox: Mailbox? = null,
    /** The tiering rung (11), OPT-IN: the READ half. The write half is the block's port. */
    val relayRead: RelayRead? = null,
    /** Per SOURCE (12.2). An unlisted source gets DurableQueue — see spine/pure/mailbox. */
    val policies: List<InputPolicy> = emptyList(),
)

/** A fully offline environment: no keys, no network, a moving clock, a fake world. */
fun offlineEnv(
    world: World = World(),
    authority: AuthorityResolver = RunAuthority(),
    clock: Clock = MovingClock(start = 1000, step = 7),
    tickets: List<Ticket> = listOf(Ticket(TicketId("4118"), "refund not received")),
    policy: ConfirmPolicy = ConfirmingAuthorities(),
    events: EventSource = ScriptedEvents(),
    verbs: List<BlockRegistration<State>>? = null,
    mailbox: Mailbox? = null,
    relayRead: RelayRead? = null,
    policies: List<InputPolicy> = emptyList(),
): Env = Env(
    clock = clock,
    authority = authority,
    oncall = world.oncall,
    delivery = world.delivery,
    relay = world.relay,
    policy = policy,
    events = events,
    tickets = tickets,
    verbs = verbs,
    mailbox = mailbox,
    relayRead = relayRead,
    policies = policies,
)

class App(
    val boundary: Boundary<State>,
    val controller: Controller<AppView>,
    val registry: Registry<State>,
    val bus: Bus,
    val sink: RecordingSink,
    val log: List<String>,
    val initial: State,
    val events: EventSource,
) {
    val state: State get() = boundary.state
    val performed: List<KeyedEffect> get() = sink.performed
}

/**
 * I2/I7: the six registrations. Each block is CONSTRUCTED here and handed a LENS onto
 * its own slice, so it never has to know what else is in State (L1).
 *
 * This is the composition root doing what a composition root is for. A block used to be
 * a loose `object` that existed whether or not anyone wired it; now the root builds each
 * one, and a block nobody constructs is a block that is not in the system.
 *
 * Declared as `get()` rather than a stored property so no top-level initialisation
 * order exists to reason about — the same idiom every slice's `empty` uses.
 */
val ALL_BLOCKS: List<BlockRegistration<State>>
    get() = listOf(
        TriageBlock().register { it.triage },
        EscalationBlock().register { it.escalation },
        ConsoleBlock().register { it.console },
        ArtifactBlock().register { it.artifact },
        AnalysisBlock().register { it.analysis },
        InboxBlock().register { it.inbox },
    )

/**
 * 11.4's "single registry, an allowlist of the agents permitted to exist", declared
 * once at the root: the HOT LOOP may recall a peer's conclusion and may NOT publish
 * one. `publishAnalysis` is not in its registry at all, so the boundary folds an
 * `Unhandled` rather than the tier relying on a promise not to call it.
 */
val FAST_TIER: List<BlockRegistration<State>>
    get() = listOf(
        TriageBlock().register { it.triage },
        EscalationBlock().register { it.escalation },
        ConsoleBlock().register { it.console },
        ArtifactBlock().register { it.artifact },
        AnalysisBlock().registerFast { it.analysis },
        InboxBlock().register { it.inbox },
    )

/** The DEEP tier does one job: think slowly, and publish what it concluded. */
val DEEP_TIER: List<BlockRegistration<State>>
    get() = listOf(
        AnalysisBlock().registerDeep { it.analysis },
        InboxBlock().register { it.inbox },
    )

fun wireApp(env: Env): App {
    val registry = registryOf(*(env.verbs ?: ALL_BLOCKS).toTypedArray())

    val log = mutableListOf<String>()
    val sink = RecordingSink(AppSink(env.oncall, env.delivery, env.relay, log))
    val initial = initialState(env.tickets)

    val boundary = Boundary(
        clock = env.clock,
        ids = env.ids,
        bus = env.bus,
        sink = sink,
        authority = env.authority,
        policy = env.policy,
        registry = registry,
        fold = ::foldApp,
        projectContext = ::projectContextApp,
        promptVersion = env.promptVersion,
        session = env.session,
        initial = initial,
    )

    val controller = Controller(
        viewOf = { projectApp(boundary.state) },
        submit = boundary::onStepFinish,
    )

    return App(boundary, controller, registry, env.bus, sink, log, initial, env.events)
}

/** I5: the verb table meets the runtime. The loop is the only file that converts. */
fun App.agentLoop(
    models: ModelProvider<LanguageModel>,
    instructions: String,
): AgentLoop<State> = AgentLoop(
    model = models.model(),
    instructions = instructions,
    registry = registry,
    stateOf = { boundary.state },
    contextOf = { boundary.context() },
    stagedOf = { listOfNotNull(events.poll()) },
    submit = boundary::onStepFinish,
)

// ── the barge-in rung, wired (12) ──────────────────────────────────────────
// The consumer reports SPINE-shaped events; the inbox block owns an app-shaped
// closed set of its own; and THIS FILE is the one place allowed to know both. That
// is L1 in one function: the spine does not name the block, the block does not name
// the consumer, and the two closed sets are joined at the root.
//
// The mapping produces ACTIONS, so a barge-in decision travels the ONE existing path
// — resolveAction → gate → fold → commit → signed Command. Nothing new is added to
// the boundary, and RunStatus is untouched.

fun consumerActions(event: ConsumerEvent): List<Action> = when (event) {
    is ConsumerEvent.Conflated -> listOf(
        Action(
            NOTE_DROP,
            rawOf(
                "source" to event.source.value,
                "reason" to "Conflated",
                "dropped" to event.dropped.toString(),
            ),
        ),
    )

    is ConsumerEvent.Duplicate -> listOf(
        Action(
            NOTE_DROP,
            rawOf("source" to event.source.value, "reason" to "Duplicate", "dropped" to "1"),
        ),
    )

    is ConsumerEvent.TurnFailed -> listOf(
        Action(NOTE_FAULT, rawOf("source" to event.source.value, "fault" to event.fault)),
    )

    is ConsumerEvent.CancelDeadlineExceeded -> listOf(
        Action(
            NOTE_FAULT,
            rawOf(
                "source" to event.source.value,
                "fault" to "abandoned: the turn ignored cancellation for ${event.afterMs}ms",
            ),
        ),
    )
}

/**
 * What a Drain commits before the consumer stops (12.3). It REQUESTS the seal rather
 * than confirming it: `confirmSeal` is irreversible and 14.3 requires a different
 * principal, so a drain cannot rubber-stamp its own finalization. The gate is not
 * suspended because the session is ending.
 */
fun drainActions(message: Message.Drain): List<Action> = listOf(Action(REQUEST_SEAL, rawOf()))

/**
 * Build the barge-in consumer — ONLY when a mailbox was supplied. An app that takes
 * neither rung pays nothing: no mailbox, no consumer, no relay read.
 */
fun wireConsumer(app: App, env: Env, runner: TurnRunner): SerialConsumer? =
    env.mailbox?.let { mailbox ->
        SerialConsumer(
            mailbox = mailbox,
            runner = runner,
            submit = app.boundary::onStepFinish,
            report = ::consumerActions,
            finalize = ::drainActions,
            policies = env.policies,
            relay = env.relayRead,
            recallSource = SourceName("analysis"),
        )
    }

/**
 * I5 once more, one level up: the TurnRunner the consumer drives.
 *
 * It builds the agent loop with THIS TURN'S OWN staged inputs, so the recall the
 * consumer already bounded is exactly what the model is shown and exactly what rides
 * the committed record. `ctx::submit` and not `boundary::onStepFinish`: the turn's
 * only channel is the revocable one.
 */
fun App.tierRunner(
    models: ModelProvider<LanguageModel>,
    instructions: String,
): TurnRunner = TurnRunner { message, ctx ->
    AgentLoop(
        model = models.model(),
        instructions = instructions,
        registry = registry,
        stateOf = { boundary.state },
        contextOf = { boundary.context(ctx.staged) },
        stagedOf = { ctx.staged },
        submit = ctx::submit,
    ).runTurn(promptFor(message))
}

/** One prompt per kind of barge-in. Closed match, no else arm. */
fun promptFor(message: Message): String = when (message) {
    is Message.Input -> message.staged.body
    is Message.Interrupt -> "INTERRUPT: ${message.reason}"
    is Message.Drain -> "DRAIN: ${message.reason}"
}

File: examples/typescript/src/app/wire.ts
// ── app/wire — the SINGLE composition root (G7) ────────────────────────
// Exactly one file may know what is real and what is faked in a build. Removing
// it means a service locator, which G7 forbids.
//
// Plugging a block in is: its `register(...)` line here, its slice field in
// app/contract, its three union memberships there, and its branch in each of
// app/assemble's three dispatchers. Pulling it out is the same list, subtracted,
// plus `rm -rf src/blocks/<X>/`. Every one of those is an APPEND to a closed
// set, and the compiler names each one you forget.

import { liveRelay } from "../blocks/analysis/adapter";
import type { AnalysisRelay } from "../blocks/analysis/register";
import { analysis } from "../blocks/analysis/register";
import { liveDelivery } from "../blocks/artifact/adapter";
import type { DeliveryPort } from "../blocks/artifact/register";
import { artifact } from "../blocks/artifact/register";
import { consoleBlock } from "../blocks/console/register";
import { livePager } from "../blocks/escalation/adapter";
import type { OncallPort } from "../blocks/escalation/register";
import { escalation } from "../blocks/escalation/register";
import { inbox } from "../blocks/inbox/register";
import { triage } from "../blocks/triage/register";
import type { Action, Registry } from "../spine/boundary/action";
import { registryOf } from "../spine/boundary/action";
import { Boundary } from "../spine/boundary/boundary";
import { InMemoryBus, sequentialIds } from "../spine/boundary/in-memory";
import type { RelayRecall, TurnRunner } from "../spine/concurrency/consumer";
import { SerialConsumer } from "../spine/concurrency/consumer";
import type { Authorization } from "../spine/ports/authorization";
import type { Bus } from "../spine/ports/bus";
import type { Clock } from "../spine/ports/clock";
import type { IdSource } from "../spine/ports/id-source";
import type { Mailbox } from "../spine/ports/mailbox";
import type { Scheduler } from "../spine/ports/scheduler";
import type { PerformMode, Sink } from "../spine/ports/sink";
import type { Actor, Authority } from "../spine/pure/actor";
import { authority } from "../spine/pure/actor";
import type { Emit } from "../spine/pure/emit";
import type { SessionId } from "../spine/pure/ids";
import type { KeyedEffect } from "../spine/pure/keyed-effect";
import type { DrainMessage, InputPolicy } from "../spine/pure/mailbox";
import type { ConsumerEvent } from "../spine/pure/turn";
import type { BlockRegistration } from "../spine/pure/verb";
import { committedSourceKeys } from "../spine/replay/replay";
import { Controller } from "../spine/surface/controller";
import { dispatchers, project } from "./assemble";
import type { AppView, Effect, State } from "./contract";
import { initialState } from "./contract";

// ── The effect sink: one branch per Effect case, exhaustive ────────────────
// A new effect kind costs TWO appends: its case in the owning block's contract,
// and one branch here. The compiler names the second one.
//
// REPLAY touches nothing (F5). RECOVERY re-drives; the deduping sink in front
// of it drops anything already acknowledged, keyed on the committed step index.

export interface Ports {
  readonly oncall: OncallPort;
  readonly delivery: DeliveryPort;
  /** the tier relay's WRITE half — the deep tier's only route to a peer (11.2) */
  readonly relay: AnalysisRelay;
  readonly log: (line: string) => void;
}

export function effectSink(ports: Ports): Sink {
  return {
    perform(keyed: KeyedEffect<Effect>, mode: PerformMode): void {
      if (mode === "REPLAY") return; // collect the descriptor; touch NOTHING
      const effect = keyed.effect;
      switch (effect.kind) {
        case "Diag":
          ports.log(`[diag @${effect.at}] ${effect.note}`);
          return;
        case "LogDecision":
          ports.log(
            `[decision @${effect.at}] ${effect.ticket} → ${effect.level}` +
              (effect.supersedes === null ? "" : ` (was ${effect.supersedes})`),
          );
          return;
        case "PageOncall":
          ports.oncall.page(effect.ticket);
          return;
        case "DeliverArtifact":
          ports.delivery.deliver(effect.lines);
          return;
        case "PublishConclusion":
          // The deep tier's write, as an ordinary effect descriptor — so REPLAY
          // stubs it and RECOVERY dedupes it on `EffectKey`, for free.
          ports.relay.publish(effect.at, effect.text);
          return;
        default: {
          const _never: never = effect;
          return _never;
        }
      }
    },
  };
}

// ── Authorization: the PRODUCT-OWNED seam (14.3, F3, F13) ──────────────────
// A real deployment resolves a principal from a session token, a policy tier's
// identity, or an approval queue's record. Here it is a supplied table, so the
// unattended-confirmer cases are exercisable offline.

export interface AuthorizationConfig {
  /** the authority this stream is acting under, per Actor */
  readonly authorities: Readonly<Record<Actor, Authority>>;
  /** the product's own rule; default-allow once the gate's structural checks
   *  (a pending request exists, and it was raised by a DIFFERENT principal)
   *  have already passed.
   *
   *  Keyed on the AUTHORITY, never the Actor — §5.2's "preserved for audit,
   *  not for branching" applies to the composition root too. A default that
   *  branched on `sig.by` here would be the exact anti-pattern G6 forbids the
   *  gate, shipped as the line every adopter copies first: it makes "a policy
   *  tier may confirm, this run may not" unrepresentable, because both are
   *  truthfully `Agent`. The Kotlin port's `ConfirmingAuthorities` keys the
   *  same way. */
  readonly mayConfirm?: (by: Authority) => boolean;
}

export function authorization(config: AuthorizationConfig): Authorization<State> {
  return {
    authorityOf: (by: Actor, _session: SessionId): Authority => config.authorities[by],
    mayConfirm: (sig) => config.mayConfirm?.(sig.authority) ?? true,
  };
}

export const defaultAuthorities: Readonly<Record<Actor, Authority>> = {
  Human: authority("host:operator"),
  Agent: authority("agent-run-7f"),
  Spine: authority("spine:consumer"),
};

// ── The application ────────────────────────────────────────────────────────

export interface Env {
  readonly clock: Clock;
  /** 11.4's allowlist: WHICH blocks this stream is permitted to run. Omitted,
   *  an app gets every block — so an app that never tiers pays nothing. */
  readonly verbs?: readonly BlockRegistration<State>[];
  readonly ids?: IdSource;
  readonly bus?: Bus;
  readonly sink: Sink;
  readonly authz?: Authorization<State>;
  readonly session?: SessionId;
  readonly promptVersion?: string;
  readonly initial?: State;
}

export interface App {
  readonly boundary: Boundary<State>;
  readonly controller: Controller<State, AppView>;
  readonly registry: Registry<State>;
  readonly bus: Bus;
  readonly dispatchers: typeof dispatchers;
  readonly initial: State;
}

// ── 11.4 — the registry allowlist, declared once at the root ───────────────
// A second tier is OPTIONAL. These three lists are the only place in the system
// that says which agents are permitted to exist; the blocks themselves know
// nothing about tiers, and neither does the spine.

export const ALL_BLOCKS: readonly BlockRegistration<State>[] = [
  triage.register<State>(),
  escalation.register<State>((s) => s.escalation),
  consoleBlock.register<State>(),
  artifact.register<State>((s) => s.artifact),
  analysis.register<State>("both"),
  inbox.register<State>(),
];

/** the hot loop: it may RECALL a peer's conclusion, never publish one */
export const FAST_TIER: readonly BlockRegistration<State>[] = [
  triage.register<State>(),
  escalation.register<State>((s) => s.escalation),
  consoleBlock.register<State>(),
  artifact.register<State>((s) => s.artifact),
  analysis.register<State>("fast"),
  inbox.register<State>(),
];

/** the deep tier: it may PUBLISH, and it holds no handle to the fast tier */
export const DEEP_TIER: readonly BlockRegistration<State>[] = [
  analysis.register<State>("deep"),
  inbox.register<State>(),
];

export function wireApp(env: Env): App {
  // ONE registration line per block. This is the plug (L1).
  const registry = registryOf<State>((env.verbs ?? ALL_BLOCKS).flatMap((r) => r.verbs));

  const bus = env.bus ?? new InMemoryBus();
  const initial = env.initial ?? initialState();
  const boundary = new Boundary<State>(
    {
      clock: env.clock,
      ids: env.ids ?? sequentialIds("cmd"),
      bus,
      sink: env.sink,
      authz: env.authz ?? authorization({ authorities: defaultAuthorities }),
      registry,
      session: env.session ?? "session-1",
      promptVersion: env.promptVersion ?? "prompt-v1",
      fold: dispatchers.fold,
      projectContext: dispatchers.projectContext,
    },
    initial,
  );

  return {
    boundary,
    controller: new Controller<State, AppView>(boundary, project),
    registry,
    bus,
    dispatchers,
    initial,
  };
}

/** The offline bindings: no keys, no network, no clients. */
export function offlinePorts(log: Emit, relay?: AnalysisRelay): Ports {
  return {
    oncall: livePager(log),
    delivery: liveDelivery(log),
    relay: relay ?? liveRelay((at, text) => log(`[relay] conclusion published @${at}: ${text}`)),
    log,
  };
}

// ── The barge-in consumer (12) — wired ONLY when a mailbox is supplied ──────
// Two mappings live here and only here, and both are L1-forced: the spine's
// `ConsumerEvent` and the inbox block's `DropReason` are separate closed sets,
// so something has to join them, and the composition root is the one place
// allowed to name both. Same for a Drain's finalization, which is the artifact
// block's business and not the consumer's.

export interface ConsumerEnv {
  readonly mailbox: Mailbox;
  readonly scheduler: Scheduler;
  /** injected, never imported — `spine/concurrency` never names the SDK */
  readonly turn: TurnRunner;
  readonly policies?: readonly InputPolicy[];
  readonly relay?: RelayRecall;
  readonly cancelDeadlineMs?: number;
  readonly drainDeadlineMs?: number;
  readonly recallDeadlineMs?: number;
}

/** ConsumerEvent → Actions. Every branch produces a real verb, so every dropped
 *  input travels the ONE existing path: resolveAction → gate → fold → commit →
 *  signed Command. Nothing the consumer sheds is silent. */
export function reportActions(event: ConsumerEvent): readonly Action[] {
  switch (event.kind) {
    case "Conflated":
      return [
        {
          tool: "noteDrop",
          input: { reason: { kind: "Conflated", source: event.source, dropped: event.dropped } },
        },
      ];
    case "Duplicate":
      return [
        {
          tool: "noteDrop",
          input: { reason: { kind: "Duplicate", source: event.source, key: event.key } },
        },
      ];
    case "TurnFailed":
      return [{ tool: "noteFault", input: { source: event.source, fault: event.fault } }];
    case "CancelDeadlineExceeded":
      return [
        {
          tool: "noteFault",
          input: {
            source: event.source,
            fault: `cancel deadline exceeded after ${event.afterMs}ms — turn abandoned, its channel revoked`,
          },
        },
      ];
    default: {
      const _never: never = event;
      return _never;
    }
  }
}

/** A Drain finalizes with the artifact block's seal REQUEST. It cannot confirm
 *  it — that needs a different principal, and the gate says so.
 *
 *  NAMED CONSEQUENCE of the spine stamp: the consumer signs its steps `Spine`, so
 *  this request is recorded under `spine:consumer` rather than `agent-run-7f`. The
 *  seal's `requestedBy` is now the spine, which makes the AGENT a legal confirmer
 *  of a drain-requested seal where it used to be the self-confirming requester the
 *  gate refused. `14.3 — the drain-requested seal and its confirmer` in
 *  test/spine/mailbox.test.ts pins that verdict, so a flip back is a red test
 *  rather than a discovery. */
export function finalizeActions(_message: DrainMessage): readonly Action[] {
  return [{ tool: "requestSeal", input: {} }];
}

export function wireConsumer(app: App, env: ConsumerEnv): SerialConsumer {
  return new SerialConsumer({
    mailbox: env.mailbox,
    scheduler: env.scheduler,
    seam: app.boundary,
    turn: env.turn,
    report: reportActions,
    finalize: finalizeActions,
    policies: env.policies,
    relay: env.relay,
    cancelDeadlineMs: env.cancelDeadlineMs,
    drainDeadlineMs: env.drainDeadlineMs,
    recallDeadlineMs: env.recallDeadlineMs,
    // Not opt-in: the dedupe scope is ALWAYS the timeline's. On a fresh bus
    // this is the empty set for free; after a crash it is what makes the
    // durable queue's redelivery refuse work that already committed (12.2).
    recovered: committedSourceKeys(app.bus.records()),
  });
}
